### PR TITLE
Add migration, model and specs for `WebVuln`

### PR DESCRIPTION
**Pull request by AI-Mozi**
#98

### CODE REVIEW

File: db/migrate/0036_create_ronin_web_vulns_table.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      foreign_key: {
+        to_table: :ronin_urls
+      }
+
+      t.string :query_param, length: 225, null: true
+      t.string :header_name, length: 225, null: true
+      t.string :cookie_param, length: 225, null: true
+      t.string :form_param, length: 225, null: true
+
+      t.string  :request_method, null: false
```
Comment (postmodern) [inline db/migrate/0036_create_ronin_web_vulns_table.rb:41]: All `enum` attributes should probably have a minimum `length:` value based on their longest possible `enum` value. It's debatable if this actually saves disk space or improves performance, but it does enforce data consistency.

**Review by postmodern — CHANGES_REQUESTED** [review]
Now that `Ronin::Vulns::SSTI#escape_type` is merged into the `0.2.0` branch, `ssti_escape` should be renamed to `ssti_escape_type`. Also `enum` columns probably need a `length:` value?

File: lib/ronin/db/web_vuln.rb [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      enum rfi_filter_bypass: {
+        null_byte: 'null_byte',
+        double_encode: 'double_encode'
+      }, _prefix: true
+
+      # @!attribute [rw] rfi_filter_bypass
+      #   The rfi filter bypass.
+      #
+      #   @return [:null_byte, :double_encode, nil]
+      enum ssti_escape: [:foo, :erb] # ! https://github.com/ronin-rb/ronin-vulns/issues/55
```
Comment (postmodern) [inline lib/ronin/db/web_vuln.rb:158]: Rename to `ssti_escape_type` and copy/paste the values from: https://github.com/ronin-rb/ronin-vulns/blob/842bffaaf10d1c427002e826c7d825c9275f11a9/lib/ronin/vulns/ssti.rb#L46

**Review by postmodern — CHANGES_REQUESTED** [review]
Also need to fill out the `ssti_escape_type` `enum` values.
https://github.com/ronin-rb/ronin-vulns/blob/842bffaaf10d1c427002e826c7d825c9275f11a9/lib/ronin/vulns/ssti.rb#L46

File: lib/ronin/db/web_vuln.rb [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      #
+      #   @return [:asp, :asp_net, :cold_fusion, :jsp, :php, :perl, nil]
+      enum rfi_script_lang: {
+        asp:         'asp',
+        asp_net:     'asp_net',
+        cold_fusion: 'cold_fusion',
+        jsp:         'jsp',
+        php:         'php',
+        perl:        'perl'
+      }
```
Comment (postmodern) [inline lib/ronin/db/web_vuln.rb:143]: This should also have a `_prefix: true`.

File: lib/ronin/db/web_vuln.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      #
+      #   @return [:double_curly_braces, :dollar_curly_braces, :dollar_double_curly_braces, :pound_curly_braces, :angle_brackets_percent, :custom, nil]
+      enum ssti_escape_type: {
+        double_curly_braces:        'double_curly_braces',
+        dollar_curly_braces:        'dollar_curly_braces',
+        dollar_double_curly_braces: 'dollar_double_curly_braces',
+        pound_curly_braces:         'pound_curly_braces',
+        angle_brackets_percent:     'angle_brackets_percent',
+        custom:                     'custom'
+      }
```
Comment (postmodern) [inline lib/ronin/db/web_vuln.rb:165]: This should also have a `_prefix: true`.

File: lib/ronin/db/web_vuln.rb [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+      #   The scheme of the URL.
+      #
+      #   @return [URL]
+      belongs_to :url, required:   true,
+                       class_name: 'URL'
+
+      # @!attribute [rw] type
+      #   The type of vuln.
+      #
+      #   @return [String, nil]
```
Comment (postmodern) [inline lib/ronin/db/web_vuln.rb:49]: `type` should be required and never `nil`.

File: lib/ronin/db/web_vuln.rb [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      enum type: {
+        lfi:           'lfi',
+        rfi:           'rfi',
+        sqli:          'sqli',
+        ssti:          'ssti',
+        open_redirect: 'open_redirect',
+        reflected_xss: 'reflected_xss'
+      }
+
```
Comment (postmodern) [inline lib/ronin/db/web_vuln.rb:58]: I think we need `validates :type, presence: true` here to ensure the `type` enum is set?

File: db/migrate/0036_create_ronin_web_vulns_table.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+#
+class CreateRoninWebVulnsTable < ActiveRecord::Migration[7.0]
+
+  def change
+    create_table :ronin_web_vulns, if_not_exists: true do |t|
+      t.string :type, length: 13, null: false
+      t.references :url, null: false,
+      foreign_key: {
+        to_table: :ronin_urls
+      }
```
Comment (postmodern) [inline db/migrate/0036_create_ronin_web_vulns_table.rb:34]: Indent.

**Review by postmodern — CHANGES_REQUESTED** [review]
Noticed a few critical issues right before I was about to merge this.

File: db/migrate/0036_create_ronin_web_vulns_table.rb [hunk 8]
```diff
@@ -50,7 +50,7 @@ def change
       t.boolean :sqli_escape_parens, null: true
       t.boolean :sqli_terminate, null: true
 
-      t.index :type, unique: true
+      t.index [:type, :query_param, :header_name, :cookie_param, :form_param, :url], unique: true, name: 'vuln_index'
```
Comment (AI-Mozi) [inline db/migrate/0036_create_ronin_web_vulns_table.rb:53]: Is `vuln_index` ok for index name?

Comment (postmodern) [reply db/migrate/0036_create_ronin_web_vulns_table.rb:53]: Probably name it `index_ronin_url_web_vulns_unique` to match the other index names.

**Review by postmodern — APPROVED** [review]

**Comment by postmodern** [comment]
Oops, this was merged into `main`. I will manually move the commit over to the `0.2.0` branch.

**Comment by postmodern** [comment]
Also `git commit --amend`ed a few missing things I forgot to mention, like adding the model class to `lib/ronin/db/models.rb` and the README.

**Pull request merged**